### PR TITLE
Paper input

### PR DESCRIPTION
**Pull request by DanChadwick**
Convert paper-input to ember-paper 1.0 standards.

- [x] Convert to `ember-paper` 1.0 coding standards.
- [x] Convert validation to computed properties.
- [x] Unify internal validations, custom validations, and external validation messages.
- [x] Support displaying multiple validations at once a la Angular Material 2.0.
- [x] Support the required asterisk a la Angular Material 2.0.
- [x] Fix the display of inputs with an icon on the left.
- [x] Change the passing of passthru attributes from `attr-` to `passThru.`
- [x] Change focus in/out events to require action closures.
- [x] Numerous improvement to the dummy app documentation page.
- [x] Allow required asterisk styling without required validation (useful for external validation messages that implement required validation). `required="style"` sets only the styling, but performs no validation. If the html5 required attribute is desired instead, use `passThru`.
- [x] Update documentation page for change to passThru attributes.
- [x] Update tests for API changes.
- [x] Document API changes in CHANGELOG.md

### CODE REVIEW

**Comment by DanChadwick** [comment]
Test failures are due to paper-dialog, not this PR.

File: addon/components/paper-input.js [hunk 1]
```diff
@@ -3,61 +3,74 @@ import BaseFocusable from './base-focusable';
...
 
+const { $, computed, isPresent, isArray, Logger, A } = Ember;
+
 export default BaseFocusable.extend(ColorMixin, FlexMixin, {
   tagName: 'md-input-container',
   classNames: ['md-default-theme'],
-  classNameBindings: ['hasValue:md-input-has-value', 'focus:md-input-focused', 'isInvalid:md-input-invalid', 'iconFloat:md-icon-float'],
+  classNameBindings: [
+    'hasValue:md-input-has-value',
+    'focus:md-input-focused',
```
Comment (miguelcobain) [inline addon/components/paper-input.js]: BaseFocusable handles this. Also, `focus` isn't used anymore. BaseFocusable uses `focused`.

File: addon/components/paper-input.js [hunk 2]
```diff
@@ -3,61 +3,74 @@ import BaseFocusable from './base-focusable';
...
+      let lineHeight = null;
+
+      if (textareaNode.hasAttribute('rows')) {
+        minRows = parseInt(textareaNode.getAttribute('rows'));
+      }
 
-    if (textareaNode.hasAttribute('rows')) {
-      minRows = parseInt(textareaNode.getAttribute('rows'));
-    }
+      textarea.on('keydown input', () => {
```
Comment (miguelcobain) [inline addon/components/paper-input.js]: Let's use event namespacing:
```js
textarea.on(`keydown.${this.elementId} input.${this.elementId}`, ...
```

File: addon/components/paper-input.js [hunk 3]
```diff
@@ -3,61 +3,74 @@ import BaseFocusable from './base-focusable';
...
-      textarea.attr('rows', '1');
+        textarea.on('scroll', () => {
+          this.onScroll(textareaNode);
+        });
+      }
 
-      textarea.on('scroll', () => {
-        this.onScroll(textareaNode);
-      });
+      $(window).on('resize', this.growTextarea(textarea, textareaNode, container, minRows, lineHeight));
```
Comment (miguelcobain) [inline addon/components/paper-input.js]: namespacing here as well

File: addon/components/paper-input.js [hunk 4]
```diff
@@ -3,61 +3,74 @@ import BaseFocusable from './base-focusable';
...
     }
+  },
 
-    Ember.$(window).on('resize', this.growTextarea(textarea, textareaNode, container, minRows, lineHeight));
+  willDestroyElement() {
+    if (this.get('textarea')) {
+      $(window).off('resize', this.growTextarea);
+      this.$().children('textarea').first().off('keydown input scroll');
```
Comment (miguelcobain) [inline addon/components/paper-input.js]: And when removing event handlers as well.
The reason for this is to make sure we don't remove unwanted event handlers. e.g another component may attach a handler to `window`, the user may be using something that attaches to `textarea`.
By using off with `event.someString` we make sure we target that specific handler.

More info: http://api.jquery.com/event.namespace/

**Comment by miguelcobain** [comment]
`icon-class` should be camelCased, correct?

File: addon/components/paper-input.js [hunk 5]
```diff
@@ -141,69 +152,72 @@ export default BaseFocusable.extend(ColorMixin, FlexMixin, {
...
-
+      });
+    } catch (error) {
+      Logger.error('Exception with custom validation: ', error);
     }
 
-    return valueIsInvalid;
-  },
+    let validationErrors = this.get('validationErrors');
+    if (validationErrors && isArray(validationErrors)) {
```
Comment (miguelcobain) [inline addon/components/paper-input.js]: `isArray` already tests that `validationErrors` is truthy.

File: app/templates/components/paper-input.hbs [hunk 6]
```diff
@@ -1,65 +1,96 @@
...
-    <div ng-messages>
-    <div id="error-{{inputElementId}}" ng-message="{{ng-message}}" class="{{if isInvalid 'ng-enter ng-enter-active' 'ng-leave ng-leave-active'}}">{{errortext}}</div>
+  <div class="md-errors-spacer">
+    {{#if maxlength}}
+      <div class="md-char-counter">{{renderCharCount}}</div>
+    {{/if}}
+  </div>
+  {{#if isInvalid}}
+    <div class="md-input-messages-animation md-auto-hide" ng-messages>
+      {{#each errorMessages as |message index|}}
```
Comment (miguelcobain) [inline app/templates/components/paper-input.hbs]: change `message` to `error`

**Comment by miguelcobain** [comment]
- Add `md-no-float` attribute when `noFloat=true`.
- Use `icon="iconName"` to include a left icon and `iconRight="iconName"` to include a right icon.
- Use `md-has-icon` class when one of them is present (`computed.or('icon','iconRight')`)
- Add `md-icon-right` class when `iconRight` is present

**Comment by DanChadwick** [comment]
All set, but note that `no-float` applies to the label, not the container or icon. It prevents the label from floating up when you start to enter text. See the revised dummy app.

**Comment by miguelcobain** [comment]
Closed in favor of https://github.com/miguelcobain/ember-paper/pull/326

**Pull request closed without merging**